src/webhook/webhook.service.ts

```diff
@@ -0,0 +1,116 @@
+    private readonly patRepository: PATRepository,
+  ) {}
+  cryptr = new Cryptr(process.env.CRYPTR_SECRET_KEY);
+
+  async createWebhook(
+    userId: number,
+    webhookData: CreateWebhookRequest,
+  ): Promise<void> {
+    const userWebhooks = await this.patRepository.getAllWebhooksByUserId(
+      userId,
+    );
+    if (userWebhooks.length >= 5) {
+      throw new ForbiddenException();
+    }
+
+    if (webhookData.token) {
+      const encryptedString = this.cryptr.encrypt(webhookData.token);
+      webhookData.token = encryptedString;
+    }
+
+    this.patRepository.createWebhook(userId, webhookData);
+  }
+
+  async deleteWebhook(userId: number, webhookId: number) {
+    const webhook = await this.patRepository.getWebhookById(webhookId);
+    if (!webhook) {
+      throw new NotFoundException();
+    }
+    if (userId !== webhook.userId) {
+      throw new UnauthorizedException();
+    }
+    this.patRepository.deleteUserWebhookByName(webhookId);
+  }
+
+  async getWebhooksById(userId: number): Promise<ListWebhooksDto[]> {
+    return this.patRepository.getAllWebhooksByUserId(userId);
+  }
+
+  sendToWebhook(url: string, data: Recipe, token?: string, attempt = 0): void {
+    let decryptedToken: string;
+    if (token) {
+      decryptedToken = this.cryptr.decrypt(token);
+    }
+    const headersRequest = {
+      'Content-Type': 'application/json',
+      Authorization: `Bearer ${token}`,
+    };
+    this.httpService
+      .post(url, data, {
+        headers: headersRequest,
+      })
+      .subscribe({
+        error: () => {
+          const maxWaitingTime = 86400 * 1000;
+          const nextTryInSec = Math.min(Math.pow(2, attempt + 1) * 1000);
+          if (nextTryInSec > maxWaitingTime) {
+            throw new NotFoundException({
+              message: 'Could not send data to provided URL',
+            });
+          }
+          setTimeout(() => {
+            this.sendToWebhook(url, data, token, attempt + 1);
+          }, nextTryInSec);
+        },
+      });
+  }
+
+  async recipeCreated(userId: number, data: Recipe) {
+    const userWebhooks = await this.patRepository.getAllWebhooksByUserId(
+      userId,
+    );
+    userWebhooks.forEach((webhook) => {
+      if (webhook.type === 'CREATE') {
+        this.sendToWebhook(webhook.url, data, webhook.token);
+      }
+    });
+  }
+
+  async recipeUpdated(userId: number, data: Recipe) {
+    const userWebhooks = await this.patRepository.getAllWebhooksByUserId(
+      userId,
+    );
+    userWebhooks.forEach((webhook) => {
+      if (webhook.type === 'UPDATE') {
+        this.sendToWebhook(webhook.url, data, webhook.token);
+      }
+    });
+  }
+
+  async recipeDeleted(userId: number, data: Recipe) {
+    const userWebhooks = await this.patRepository.getAllWebhooksByUserId(
+      userId,
+    );
+    userWebhooks.forEach((webhook) => {
+      if (webhook.type === 'DELETE') {
+        this.sendToWebhook(webhook.url, data, webhook.token);
+      }
+    });
+  }
```

Comment:
hmm, I think it's a bit too much redundancy and we are breaking the DRY principle here (you have 3x almost the same code repeated in the same file). I'd do it differently:

- have one method that sends webhook events - in your case `sendToWebhook`.
- I'd create an enum for events like this:

```typescript
enum WebhookEvent {
  RecipeCreated = "recipe_created",
  RecipeUpdated = "recipe_updated",
  RecipeDeleted = "recipe_deleted"
}
```
- `sendToWebhook` method should accept `WebhookEvent` as parameter
- looking at the implementation we would need an additional method - e.g. `sendWebhookEvent`, where we will include this part: `userWebhooks.forEach((webhook) => { ... })`

Reply:
A word on naming the events:

Please note that this feature is something that our customers integrate with. We have no control over these integrations. Thus we have to be careful when making any changes, as it might break customer's integration. When we really have to make a breaking change, we have to create a `v2` webhook, implement the possibility of letting user decide wich version to use, deprecate `v1` webhook and give users time to migrate to `v2`. All of this takes a lot of time.

Here we should immediately change event names - we have now `CREATE`, `UPDATE` and `DELETE`. What if we want to also have events for when a new user was created? We would create (I persume) a `USER_CREATED` event. But now it would be good to change `CREATE` to `RECIPE_CREATED` as otherwise it might be confusing. But if users are already using our webhooks - we cannot do this easily.

Reply:
I understand your idea about this implementation, but shouldn't `sendWebhookEvent` accept parameter `WebhookEvent`? I have to specify in `recipe.service` to which type of webhook event should be send, therefore `sendWebhookEvent` should process it and then `sendToWebhook` just sends it.

Reply:
yes, `sendWebhookEvent` should accept `WebhookEvent` as parameter 👍🏻